### PR TITLE
additional tests to improve coverage

### PR DESCRIPTION
**Pull request by bkamins**
No functionality is changed in this PR

### CODE REVIEW

**Comment by bkamins** [comment]
additional coverage improvement is in https://github.com/JuliaData/DataFrames.jl/pull/2554. I will come back to that PR tomorrow to finalize it.

**Comment by bkamins** [comment]
@nalimilan - this should be good for a review

File: test/broadcasting.jl [hunk 1]
```diff
@@ -133,6 +136,12 @@ end
 end
 
 @testset "normal data frame and data frame row in broadcasted assignment - one column" begin
+    df = DataFrame(a=1, b=2)
+    @test_throws ArgumentError df .= DataFrame(b=1, a=2)
+    @test_throws ArgumentError df .= DataFrame(a=1, c=2)
+    @test_throws ArgumentError df[!, [:a, :b]] .= DataFrame(b=1, a=2)
+    @test_throws ArgumentError df[!, [:a, :b]] .= DataFrame(a=1, c=2)
```
Comment (nalimilan) [inline test/broadcasting.jl:143]: This testset seems to be about single columns, right? Better put this elsewhere?

File: test/broadcasting.jl [hunk 2]
```diff
@@ -51,6 +51,9 @@ end
     @test (df1 .- df2) == DataFrame(zeros(size(refdf)), names(refdf))
     @test (df1 .* df2) == refdf .^ 2
     @test (df1 ./ df2) == DataFrame(ones(size(refdf)), names(refdf))
+
+    @test_throws ArgumentError DataFrame(a=1, b=1) .+ DataFrame(b=1, a=1)
+    @test DataFrame() .+ DataFrame() == DataFrame()
```
Comment (nalimilan) [inline test/broadcasting.jl:56]: This testset seems to be about non-`DataFrame` arguments, so better put this elsewhere?

File: test/constructors.jl [hunk 3]
```diff
@@ -163,6 +166,11 @@ end
     @test isequal(df, DataFrame(x1 = [0.0, 0.0, 0.0], x2 = [1.0, 1.0, 1.0]))
     @test isapprox(df, DataFrame(x1 = [0.0, 0.0, 0.0], x2 = [1.0, 1.0, 1.0]))
     @test isapprox(df, DataFrame(x1 = [0.0, 0.0, 0.0], x2 = [1.000000010000, 1.0, 1.0]))
+    @test_throws DimensionMismatch isapprox(DataFrame(a=1), DataFrame(a=[1,2]))
+    @test_throws ArgumentError isapprox(DataFrame(a=1), DataFrame(b=1))
+    @test !isapprox(df, DataFrame(x1 = [0.0, 0.0, 0.0], x2 = [1.1, 1.0, 1.0]))
+    @test !isapprox(df, DataFrame(x1 = [0.0, 0.0, 0.0], x2 = [1.1, 1.0, 1.0]), atol=0.09)
+    @test isapprox(df, DataFrame(x1 = [0.0, 0.0, 0.0], x2 = [1.1, 1.0, 1.0]), atol=0.11)
```
Comment (nalimilan) [inline test/constructors.jl:173]: Here too, the testset is about the pair constructor, which isn't used here.

File: test/dataframerow.jl [hunk 4]
```diff
@@ -48,6 +50,7 @@ end
     sdf = view(df, [5, 3], [3, 1, 2])
 
     r = DataFrameRow(df, 2, :)
+    @test_throws ArgumentError r .+ 1
```
Comment (nalimilan) [inline test/dataframerow.jl:53]: And here the testset is about indexing.

File: test/dataframerow.jl [hunk 5]
```diff
@@ -121,6 +124,7 @@ end
     df = deepcopy(ref_df)
     df2 = DataFrame(a = [1, 2, 3])
 
+    @test !isequal(DataFrameRow(df, 1, :), DataFrameRow(df2, 1, :))
```
Comment (nalimilan) [inline test/dataframerow.jl:127]: Maybe also test a case where it's true (including with a missing value)?

File: test/grouping.jl [hunk 6]
```diff
@@ -3224,4 +3230,47 @@ end
...
+    @test select(gdf, [:x], :id => (x -> fill([1], length(x))) => [:x]) ==
+          DataFrame(id=df.id, x=1)
+    @test_throws ArgumentError select(gdf, x -> [1, 2])
+
+    df = DataFrame(id=[1, 1, 2, 3, 3, 1], x=categorical(1:6, ordered=true))
+    gdf = groupby_checked(df, :id)
+    @test combine(gdf, :x => minimum => :x) == df[[1, 3, 4], :]
+end
+
+@testset "select and transform! tests with leading function" begin
```
Comment (nalimilan) [inline test/grouping.jl:3259]:
```suggestion
@testset "select and transform! tests with function as first argument" begin
```

File: test/grouping.jl [hunk 7]
```diff
@@ -3224,4 +3230,47 @@ end
...
+    select!(sdf -> sdf.id .* sdf.x, gdf)
+    @test df == df2
+
+    df = DataFrame(id=[1, 1, 2, 3, 3, 1], x=1:6)
+    gdf = groupby_checked(df, :id)
+    df2 = transform(sdf -> sdf.id .* sdf.x, gdf)
+    @test df2 == DataFrame(id=df.id, x=df.x, x1=df.id .* df.x)
+    transform!(sdf -> sdf.id .* sdf.x, gdf)
+    @test df == df2
+
```
Comment (nalimilan) [inline test/grouping.jl:3273]:
```suggestion
```

File: test/reshape.jl [hunk 8]
```diff
@@ -647,4 +647,12 @@ end
     @test permutedims(DataFrame(a=String[], b=Float64[]), 1) == DataFrame(a=["b"])
 end
 
+@testset "stack view=true additional tests" begin
+    df = DataFrame(a=1:3, b=11:13, c=101:103)
+    sdf = stack(df, [:b, :c], view=true)
+    @test reverse(sdf.a) == reverse(copy(sdf.a))
+    @test IndexStyle(DataFrames.StackedVector) == IndexLinear()
+    sdf.value
```
Comment (nalimilan) [inline test/reshape.jl:655]: What's this?

**Review by nalimilan — COMMENTED** [review]
Thanks! I've just spotted a few tests that seem a bit out of place. It's always hard to find out to organize tests...

File: test/reshape.jl (hunk 8 above)
Comment (bkamins) [reply test/reshape.jl:655]: fixed

File: test/broadcasting.jl (hunk 2 above)
Comment (bkamins) [reply test/broadcasting.jl:56]: ok - moved

File: test/broadcasting.jl (hunk 1 above)
Comment (bkamins) [reply test/broadcasting.jl:143]: also moved

File: test/constructors.jl (hunk 3 above)
Comment (bkamins) [reply test/constructors.jl:173]: OK - I moved the whole `isapprox` part, as it was the only place where earlier we tested `isapprox`.

File: test/dataframerow.jl (hunk 4 above)
Comment (bkamins) [reply test/dataframerow.jl:53]: ok - added a new testset

File: test/dataframerow.jl (hunk 5 above)
Comment (bkamins) [reply test/dataframerow.jl:127]: added

**Comment by bkamins** [comment]
updated

**Review by nalimilan — APPROVED** [review]
98.24% coverage! Maybe we can aim at 100% one day! :-)

**Comment by bkamins** [comment]
> Maybe we can aim at 100% one day! :-)

No 😄, because:
1. there are some lines that CI skips (e.g. `local variable_name` statement)
2. there are some lines that should be never executed (error messages in code paths that should not be normally reachable)

But we should be close.

Thank you for reviewing this.

**Pull request merged**